src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1490 @@
+     */
+    @SuppressWarnings("serial")
+    private static class RefSet<T extends Ref> extends HashSet<T> {
+
+        public static <T extends Ref> RefSet<T> newEmpty() {
+            return new RefSet<>();
+        }
+
+        /**
+         * Discard any {@link ExprRef}'s at the specified depth.
+         * Do this when discarding whatever is on top of the stack.
+         */
+        public boolean discardExprs(int depth) {
+            return remove(ExprRef.direct(depth)) | remove(ExprRef.indirect(depth));
+        }
+
+        /**
+         * Extract any {@link ExprRef}'s at the specified depth.
+         */
+        public RefSet<ExprRef> removeExprs(int depth) {
+            return Stream.of(ExprRef.direct(depth), ExprRef.indirect(depth))
+              .filter(this::remove)
+              .collect(Collectors.toCollection(RefSet::new));
+        }
+
+        /**
+         * Extract any {@link ExprRef}'s at the specified depth and do something with them.
+         */
+        public void removeExprs(int depth, Consumer<? super Boolean> handler) {
+            Stream.of(ExprRef.direct(depth), ExprRef.indirect(depth))
+              .filter(this::remove)
+              .map(ExprRef::isDirect)
+              .forEach(handler);
+        }
+
+        /**
+         * Replace any references of the given type.
+         */
+        public void replace(Class<? extends Ref> type, Function<Boolean, ? extends T> mapper) {
+            final List<Ref> oldRefs = this.stream()
+              .filter(type::isInstance)
+              .collect(List.collector());             // avoid ConcurrentModificationException
+            this.removeAll(oldRefs);
+            oldRefs.stream()
+              .map(Ref::isDirect)
+              .map(mapper)
+              .forEach(this::add);
+        }
+
+        /**
+         * Replace any {@link ExprRef}'s at the specified depth.
+         */
+        public void replaceExprs(int depth, Function<Boolean, ? extends T> mapper) {
+            removeExprs(depth, direct -> add(mapper.apply(direct)));
+        }
+
+        /**
+         * Find references of the given type, map them, and add them to {@code dest}.
+         */
+        public <S extends Ref> void mapInto(RefSet<S> dest, Class<? extends Ref> type,
+                Function<Boolean, ? extends S> mapper) {
+            final List<S> newRefs = this.stream()
+              .filter(type::isInstance)
+              .map(Ref::isDirect)
+              .map(mapper)
+              .collect(List.collector());             // avoid ConcurrentModificationException
+            dest.addAll(newRefs);
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public RefSet<T> clone() {
+            return (RefSet<T>)super.clone();
+        }
+    }
+
+// MethodInfo
+
+    // Information about a constructor or method in the compilation unit
+    private static class MethodInfo {
```

Comment:
This could be a record, right? I believe you can then drop accessors and toString?

Reply:
Yes! Will fix - thanks.